Add Workouts extension

samuelkraft opened the pull request:
## Description

This PR adds the Workouts Extension, allowing you to view your workouts, See your totals in the menubar and ask the Workout AI about your history.

Currently only Strava is supported and Strava API doesn't allow you to name your app with their brand hence the Workouts naming + it allows adding more providers in the future.

## Screencast

https://github.com/raycast/extensions/assets/2223418/fce1a54e-79f8-4948-901d-d45a7c555343

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

pernielsentikaer reviewed (changes requested):
Hi 👋 

Thanks for your contribution 💪 

I have now tested your extension, and I have some feedback ready for you:

- Should we call the extension `Strava Workouts` instead?

- In the menubar it says 0.0km running, but it looks like I have some running this year

![Raycast26042024-iZpUDllZ png@2x](https://github.com/raycast/extensions/assets/38033107/7937cae8-f0f4-4c6c-9d41-99dcfe3ea1e3)

- When I click on one of the items in the menubar I end up here: https://www.strava.com/athletes/undefined

- I would like to see stats for a week and maybe more sport types in the menubar

- Maybe set some of the preferences a required so users knows it's possible to set, not sure if that makes sense?

- When I asked Workout AI I got this error

![workouts 2024-04-26 at 11 38 54](https://github.com/raycast/extensions/assets/38033107/acfb26da-2a54-49c8-bdeb-a56df9715448)

```
11:39:38.620 TypeError: Cannot read properties of undefined (reading 'substring')
    at Ask (/Users/pernielsentikaer/.config/raycast/extensions/workouts/ask.js:14008:81)
    at Wl (/Applications/Raycast.app/Contents/Resources/RaycastNodeExtensions_RaycastNodeExtensions.bundle/Contents/Resources/api/node_modules/react-reconciler/react-reconciler.production.min.js:5:21246)
    at ho (/Applications/Raycast.app/Contents/Resources/RaycastNodeExtensions_RaycastNodeExtensions.bundle/Contents/Resources/api/node_modules/react-reconciler/react-reconciler.production.min.js:7:41192)
    at co (/Applications/Raycast.app/Contents/Resources/RaycastNodeExtensions_RaycastNodeExtensions.bundle/Contents/Resources/api/node_modules/react-reconciler/react-reconciler.production.min.js:7:36991)
    at Ds (/Applications/Raycast.app/Contents/Resources/RaycastNodeExtensions_RaycastNodeExtensions.bundle/Contents/Resources/api/node_modules/react-reconciler/react-reconciler.production.min.js:7:36922)
    at qr (/Applications/Raycast.app/Contents/Resources/RaycastNodeExtensions_RaycastNodeExtensions.bundle/Contents/Resources/api/node_modules/react-reconciler/react-reconciler.production.min.js:7:36779)
    at vu (/Applications/Raycast.app/Contents/Resources/RaycastNodeExtensions_RaycastNodeExtensions.bundle/Contents/Resources/api/node_modules/react-reconciler/react-reconciler.production.min.js:7:33275)
    at io (/Applications/Raycast.app/Contents/Resources/RaycastNodeExtensions_RaycastNodeExtensions.bundle/Contents/Resources/api/node_modules/react-reconciler/react-reconciler.production.min.js:7:32231)
    at Lu (/Applications/Raycast.app/Contents/Resources/RaycastNodeExtensions_RaycastNodeExtensions.bundle/Contents/Resources/api/node_modules/react-reconciler/react-reconciler.production.min.js:1:1694)
    at Immediate.zu [as _onImmediate] (/Applications/Raycast.app/Contents/Resources/RaycastNodeExtensions_RaycastNodeExtensions.bundle/Contents/Resources/api/node_modules/react-reconciler/react-reconciler.production.min.js:1:2088)
11:39:38.684 
TypeError: Cannot read properties of undefined (reading 'substring')

Ask:ask.tsx:8:55

---
 5: import { getPreferenceValues } from "@raycast/api";
 6: 
 7: function Ask(props: { arguments: Arguments.Ask }) {
 8:   const { data: activities, error } = useCachedPromise(() => getActivities(1, 100));
 9: 
10:   useEffect(() => {
11:     if (error) {
---

    at withAccessToken(Ask)
Ir:index.js:11:2490
    at ray-navigation-stack
```

I'm looking forward to testing this extension again 🔥 

Request a new review when you are ready. Feel free to contact me here or at [Slack](https://www.raycast.com/community) if you have any questions.

samuelkraft commented:
> * Should we call the extension `Strava Workouts` instead?

Unfortunately this isn't allowed by the Strava Brand Guidelines, but I did add Strava as tags and made the icon "somewhat" close
![Sam 2024-04-26 at 11  48 27@2x](https://github.com/raycast/extensions/assets/2223418/38fa7e1f-5471-4ce2-824d-0cdc4b428950)

> * In the menubar it says 0.0km running, but it looks like I have some running this year

The menubar defaults to last 4 weeks (configurable in settings), in which you haven't ran anything. I pushed an update that now always displays the mileage in the dropdown for the "primary" sport even if no mileage. I think this + showing the settings on app launch (replied below) will help with the confusion.

![Sam 2024-04-26 at 12  17 24@2x](https://github.com/raycast/extensions/assets/2223418/bb64240b-f8df-4319-89e3-5364baf90360)

> * When I click on one of the items in the menubar I end up here: https://www.strava.com/athletes/undefined

Thanks for the notice, fixed! ✅

> * I would like to see stats for a week and maybe more sport types in the menubar

Great feature suggestion! The current implementation uses Stravas [stats api](https://developers.strava.com/docs/reference/#api-Athletes-getStats) which is quite limited only returns these timeframes and sports. We could theoretically fetch activities instead and calculate the numbers but this likely only works well for the last week or so for people that uploads a lot of activities. Let's add weekly distance in a follow up PR if more people request this?

> * Maybe set some of the preferences a required so users knows it's possible to set, not sure if that makes sense?

Added km/mi setting as required ✅


> * When I asked Workout AI I got this error

This should be fixed now, could you please try again?

pernielsentikaer commented:
Unfortunately this isn't allowed by the Strava Brand Guidelines, but I did add Strava as tags and made the icon "somewhat" close

> That makes sense 🔥

The menubar defaults to last 4 weeks (configurable in settings), in which you haven't ran anything. I pushed an update that now always displays the mileage in the dropdown for the "primary" sport even if no mileage. I think this + showing the settings on app launch (replied below) will help with the confusion.

> Love it, it makes it easy to understand 👏

Thanks for the notice, fixed! ✅

> Awesome, works much better 👏 

This should be fixed now, could you please try again?

> It went a bit longer, I got this one

![workouts 2024-04-26 at 12 25 10](https://github.com/raycast/extensions/assets/38033107/70edbfbc-a6d5-4069-9aa2-57013d091768)

After I got above error it's just loading forever without any response, do you have a clue what that might be because?
